### PR TITLE
Keep display discovery responsive and reuse layouts on new hardware

### PR DESCRIPTION
**Pull request by mapleroyal**
Hey, I worked with Astra Ultra on this while using your plugin over the last week. These are its comments:

This keeps unfamiliar monitors from being disabled by the wrong saved profile, lets clients reuse a layout on different hardware, and bounds the daemon's display discovery reads.

- Only select a profile automatically if it accounts for every connected screen. Keep the existing behavior for undocking and manual profile selection.
- Add `reuse_profile` and a typed IPC client helper so an editor can map a saved layout onto current screens and receive a new draft without applying or saving it.
- Remap mirrors and workspaces, retain extra connected screens, and report adjustments to unsupported modes or scales. Invalid mirror dependencies become independent displays with a warning. Leave saved command hooks out of the new draft.
- Recover unreported calibration from a separate current exact/best profile by current output key, including unassigned outputs. Only use the selected template as a calibration source for a unique matching serial, preventing uncertain hardware from inheriting the template's old settings.
- Bound each daemon monitor/workspace query to 750 ms. Return a typed `compositor_busy` error and keep a busy daemon connected in the terminal editor.
- Include an additive hardware snapshot hash in status, editor, and reuse responses. Verify hardware before and after reading workspace rules, rejecting changed bindings so clients can correlate drafts with current status.
- Skip unnecessary graphics-driver identity queries and coalesce blocked probes. Topology probes run off the event loop, which can consume hotplug/suspend events while those probes run; discard outdated probe results and retry transient failures without requiring another monitor event.
- Add timing logs for monitor events, queries, and reconciliation to help locate remaining docking delays.

The [companion Omarchy panel PR](https://github.com/crmne/omarchy-hyprmoncfg/pull/18) adds the layout-reuse form and recovery behavior.

Local validation: full Go tests, vet, and both command builds pass. Race tests pass for daemon, profile, appstatus, IPC, Hyprland client, and TUI packages. `go mod tidy` leaves module files unchanged. Regression coverage includes calibration preservation and ambiguous-template rejection, snapshot changes during a query, mirror repair, typed reuse/error responses, and busy-to-ready daemon connections.

The 750 ms deadline is per query. Status/editor/reuse snapshots perform two monitor reads and one workspace-rule read. Automatic reconciliation and profile apply/rollback retain their serialized lifecycle, so event consumption can wait during reconciliation. The terminal editor's separate asynchronous direct refresh still has its existing eight-second overall context; the Omarchy panel uses bounded daemon IPC. A physical unfamiliar-dock replug has not been tested.

### CODE REVIEW

File: internal/profile/reuse.go [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+			warnings = append(warnings, fmt.Sprintf("%s: adjusted scale to %.5g for its display mode.", current.Name, adjusted))
+		}
+		// Color/ICC and hardware overrides are display-specific. Keep live
+		// settings when identity changes, even for another unit of one model.
+		if source.MatchIdentity() != current.MatchIdentity() || strings.TrimSpace(source.Serial) == "" || counts[current.MatchIdentity()] > 1 || savedIdentityCounts[source.MatchIdentity()] > 1 {
+			output.VRR, output.Bitdepth, output.CM = current.VRR, current.Bitdepth, current.CM
+			output.SDRBrightness, output.SDRSaturation = current.SDRBrightness, current.SDRSaturation
+			output.SDRMinLuminance, output.SDRMaxLuminance = current.SDRMinLuminance, current.SDRMaxLuminance
+			output.MinLuminance, output.MaxLuminance, output.MaxAvgLuminance = current.MinLuminance, current.MaxLuminance, current.MaxAvgLuminance
+			output.SupportsWideColor, output.SupportsHDR = current.SupportsWideColor, current.SupportsHDR
```
[the remark this answers is not in the document]
Comment (mapleroyal) [reply internal/profile/reuse.go]: Addressed in [301bd1a](https://github.com/mapleroyal/hyprmoncfg/commit/301bd1a57d852514cf1741c1f84a69ccf5edbc25). Reuse now restores unreported settings from a separate current exact/best profile by current output key, including unassigned displays. The selected template supplies fallback calibration only for a unique matching serial, so serialless/duplicate-serial hardware cannot reintroduce the template's old settings. Regression tests cover mapped/unassigned outputs, swapped/skipped roles, ambiguous identities, and different serials on the same connector. The original saved profiles remain intact.

File: internal/ipc/protocol.go [hunk 2]
```diff
@@ -29,6 +29,7 @@ const (
 	MethodSubscribe   = "subscribe"
 	MethodEditor      = "editor_state"
 	MethodEdit        = "edit_profile"
+	MethodReuse       = "reuse_profile"
```
[the remark this answers is not in the document]
Comment (mapleroyal) [reply internal/ipc/protocol.go:32]: Addressed in [301bd1a](https://github.com/mapleroyal/hyprmoncfg/commit/301bd1a57d852514cf1741c1f84a69ccf5edbc25). Added `Client.ReuseProfile` with a client/server round-trip test covering the selected name, explicit skipped roles, returned draft/warnings, and retryable errors.

File: internal/ipc/server.go [hunk 3]
```diff
@@ -267,6 +279,9 @@ func decodeParams(raw json.RawMessage, target any) error {
 }
 
 func encodeResponseError(err error) *ResponseError {
+	if errors.Is(err, ErrCompositorBusy) {
+		return &ResponseError{Code: "compositor_busy", Message: ErrCompositorBusy.Error()}
```
[the remark this answers is not in the document]
Comment (mapleroyal) [reply internal/ipc/server.go:283]: Addressed in [301bd1a](https://github.com/mapleroyal/hyprmoncfg/commit/301bd1a57d852514cf1741c1f84a69ccf5edbc25). The client now preserves `ErrCompositorBusy` identity. The terminal editor also treats this response as an unknown/retryable daemon state and retains a newly connected daemon that replies busy. Real socket tests cover busy-to-ready recovery on the same connection and reconnecting to a busy daemon.

File: internal/profile/reuse.go [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+		idx := byTarget[targetKey]
+		mirrorKey := mapping[source.MirrorOf]
+		if mirrorKey == "" {
+			warnings = append(warnings, fmt.Sprintf("%s: skipped mirror source; retained it as an independent display.", draft.Outputs[idx].Name))
+			continue
+		}
+		mirror := draft.Outputs[byTarget[mirrorKey]]
+		if !mirror.Enabled {
+			return Profile{}, nil, fmt.Errorf("the mirror source for %s must be enabled", draft.Outputs[idx].Name)
+		}
```
[the remark this answers is not in the document]
Comment (mapleroyal) [reply internal/profile/reuse.go:139]: Addressed in [301bd1a](https://github.com/mapleroyal/hyprmoncfg/commit/301bd1a57d852514cf1741c1f84a69ccf5edbc25). A mapped output whose mirror source is disabled now becomes independent with a warning. Existing validation still requires an enabled independent display. Added a regression using a saved profile with that dependency.

**Comment by mapleroyal** [comment]
Pushed [301bd1a](https://github.com/mapleroyal/hyprmoncfg/commit/301bd1a57d852514cf1741c1f84a69ccf5edbc25) addressing all four inline findings: calibration recovery, the typed reuse helper, retryable busy-error handling (including terminal-editor connection behavior), and disabled-mirror repair. It also adds shared hardware snapshot hashes and verifies the hardware bindings around workspace reads for the companion panel's response correlation.

For the two suppressed timing findings, the code paths are real pre-existing limitations. The documentation and PR description now make their scope explicit: topology probes run off the event loop, while reconciliation/apply remain serialized; the terminal editor's separate asynchronous direct refresh still uses an eight-second overall context. Moving the entire apply lifecycle or redesigning terminal refresh is outside this change.

Validation: full Go tests, vet, both command builds, unchanged module files after `go mod tidy`, and race tests for daemon/profile/appstatus/IPC/Hyprland-client/TUI packages pass. Independent calibration checks included eight additional serialless/duplicate-serial and swapped/skipped-assignment cases with the selected template present in the donor list. No remaining finding from that review; no physical dock replug was performed.

**Comment by mapleroyal** [comment]
Followed up on [this review](https://github.com/crmne/hyprmoncfg/pull/61#pullrequestreview-5275579605) in [fc62fe6](https://github.com/mapleroyal/hyprmoncfg/commit/fc62fe699e4034c997306ec3eca70737c1d8c44c), separating the project import from standard-library imports for consistency.

One correction to the review: the original file already produced an empty `gofmt -d`; gofmt does not automatically separate standard-library and project imports. This is a style cleanup, with no runtime change.

I also rechecked the four previously resolved findings and their regression coverage. Full Go tests, vet, both command builds, and race tests across the affected packages pass. Module files remain unchanged after `go mod tidy`. No physical dock replug was performed for this follow-up.

— Astra Ulta

File: internal/daemon/daemon.go [hunk 5]
```diff
@@ -409,10 +474,16 @@ func (s *Service) Run(ctx context.Context) error {
 			stopDebounce()
 			debounceTimer.Reset(next.delay)
 		case <-debounceTimer.C:
-			if !pending {
+			if systemSuspended || !pending {
+				continue
+			}
+			err := s.tryApplyBest(ctx)
```
[the remark this answers is not in the document]
Comment (mapleroyal) [reply internal/daemon/daemon.go]: Agreed: the apply engine's direct reads bypassed the daemon's query deadline, and its validation timer could not interrupt a blocked monitor read. Fixed in the [updated branch](https://github.com/mapleroyal/hyprmoncfg/commit/dcf342052ed4258ec3b67a36ab1004b67c90075e) with separate per-read deadlines, retryable timeout errors, and regressions covering bounded failure, config restoration, and successful retry. Version/capability, workspace, validation-monitor, and Lua-verification reads are covered.

The apply/rollback lifecycle remains serialized. This bounds compositor reads rather than imposing a blanket timeout on display writes or post-apply commands. Full Go tests, vet, both builds, and affected-package race tests pass; module files remain unchanged after tidy.

— Astra Ulta

**Comment by mapleroyal** [comment]
Addressed [this review](https://github.com/crmne/hyprmoncfg/pull/61#pullrequestreview-5281172282) and rebased the contribution onto current upstream main (`e167f41`), resolving the conflicts.

In addition to the apply-read deadlines discussed inline, repaired mirror dependencies now receive nonoverlapping independent geometry with warnings. Regression cases cover skipped sources, disabled sources, and mirror chains.

The recommendation concern was valid against the earlier branch. Current upstream has since introduced automatic extension and explicit strict-display policy, so the rebased contribution preserves that behavior and keeps daemon/status/TUI selection consistent. It no longer imposes the earlier complete-display-only automatic filter. The PR description has been updated accordingly.

Full tests, vet, both command builds, and race tests for apply, daemon, profile, appstatus, IPC, Hyprland client, and TUI pass. Tidy leaves module files unchanged. No physical unfamiliar-dock replug was performed.

— Astra Ulta

File: internal/daemon/daemon.go [hunk 6]
```diff
@@ -478,10 +538,16 @@ func (s *Service) Run(ctx context.Context) error {
...
```
[the remark this answers is not in the document]
Comment (mapleroyal) [reply internal/daemon/daemon.go]: Fixed in [8ef6980](https://github.com/mapleroyal/hyprmoncfg/commit/8ef6980c4503ab374ac68190d3028e4ccf7ce10b). A consumed hotplug now cancels earlier debounce/retry timers and defers competing triggers until the latest topology probe completes. That result then starts a fresh debounce; a timed-out latest probe retains the retry obligation, including when a poll replaced its queued request.

Regression tests reproduce premature reconciliation with a debounce shorter than the query timeout, and cover a competing lid trigger plus timeout recovery without another hotplug event. An independent ordering audit also checked the timer/result boundary. Existing in-progress applies remain serialized.

— Astra Ulta

File: internal/profile/reuse.go [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+	for _, candidate := range profiles {
+		if candidate.Name != saved.Name {
+			donors = append(donors, candidate)
+		}
+	}
+	if active, ok := ExactStateMatch(donors, monitors, rules); ok {
+		PreserveUnreportedSettings(&draft, active)
+	} else if best, _, ok := BestMatch(donors, monitors); ok {
+		PreserveUnreportedSettings(&draft, best)
+	}
```
[the remark this answers is not in the document]
Comment (mapleroyal) [reply internal/profile/reuse.go]: Fixed in [ca27240](https://github.com/mapleroyal/hyprmoncfg/commit/ca272407d961beef91e8b02f2d18474a27a9c57d). After recovering target-device calibration, reuse explicitly restores the selected template's `DisableUnknownOutputs` policy.

Regression tests cover both policy values with no donor, an exact donor, and a best-match donor. They also verify that donor calibration is retained and the saved template is unchanged.

— Astra Ulta

File: internal/profile/reuse.go [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+	// Unassigned outputs stay live. Outputs made independent by mirror repair
+	// also need their own space, even when assigned a saved role. Preserve all
+	// other mapped geometry and move only overlapping extras/repaired mirrors.
+	for idx := range draft.Outputs {
+		output := &draft.Outputs[idx]
+		if used[output.Key] && !repairedMirrors[output.Key] {
+			continue
+		}
+		if !used[output.Key] {
+			warnings = append(warnings, fmt.Sprintf("%s was not assigned a saved role; kept its current settings.", output.Name))
```
[the remark this answers is not in the document]
Comment (mapleroyal) [reply internal/profile/reuse.go]: Fixed in [ca27240](https://github.com/mapleroyal/hyprmoncfg/commit/ca272407d961beef91e8b02f2d18474a27a9c57d). A mode or sharp-scale adjustment that changes a mapped output's logical rectangle now makes that output eligible for collision repair. Unaffected mapped positions are preserved, moved outputs produce an adjustment warning, and valid mirrors follow their relocated source. Final layout validation rejects any remaining overlap before returning the draft.

Regression tests cover mode fallback, rotated output geometry, sharp-scale adjustment, mirror alignment, and a pre-existing invalid overlap.

— Astra Ulta

**Comment by mapleroyal** [comment]
Fixed two more edge cases in 18893d9 and a3ab98e: an old display check could delay lid-open recovery, and reusing a layout could change workspace order. Both now have regression tests; local checks and CI pass.

The summary says six issues remain but lists none. The earlier formatting nit is already fixed. Please point to a specific remaining failure so we can assess it.

— Astra Ulta

**Comment by mapleroyal** [comment]
@crmne Could you share the four specific findings from [the latest review](https://github.com/crmne/hyprmoncfg/pull/61#pullrequestreview-5285910946)? It reviewed a3ab98e, but says "Findings: None" and contains no inline comments. All eight earlier review threads are resolved, and CI passes.

We can address concrete bugs in this PR, but need the actual findings before making further changes.

Astra Ulta

File: internal/profile/reuse.go [hunk 9]
```diff
@@ -0,0 +1,295 @@
...
+		if output.MirrorOf != "" {
+			source := draft.Outputs[byTarget[output.MirrorOf]]
+			output.X, output.Y = source.X, source.Y
+		}
+	}
+	enabled := 0
+	for _, output := range draft.Outputs {
+		if output.Enabled && output.MirrorOf == "" {
+			enabled++
+		}
```
[the remark this answers is not in the document]
Comment (mapleroyal) [reply internal/profile/reuse.go]: Fixed in [b549f77](https://github.com/mapleroyal/hyprmoncfg/commit/b549f7796a89d94a052dc90bff854732c11b7693). Skipping every saved role could leave only a zero-size output, so reuse now requires an enabled independent real display with positive draft dimensions. Disabled or modeless targets remain supported when an advertised mode can configure them.

`FALLBACK` was already filtered by the monitor client before IPC; the shared reuse helper now also excludes it from this check. Regression tests, full Go tests, vet, both command builds, and profile race tests pass. Module files remain unchanged after tidy.

Astra Ulta

**Comment by crmne** [comment]
The display discovery work from this PR shipped in 1.19.0 as 1d634ab, under your name. I took it out of this PR so it's focused on layout reuse; more in https://github.com/crmne/omarchy-hyprmoncfg/pull/18#issuecomment-5822126667
